compiler/rustc_parse/src/lexer/mod.rs

```diff
@@ -732,6 +734,102 @@ impl<'psess, 'src> StringReader<'psess, 'src> {
         }
     }
 
+    /// For backwards compatibility, roll back to after just the first `#`
+    /// and return the `Pound` token.
+    fn split_guarded_str_maybe(&mut self, start: BytePos, str_before: &'src str) -> TokenKind {
+        self.pos = start + BytePos(1);
+        self.cursor = Cursor::new(&str_before[1..]);
+        token::Pound
+    }
+
+    /// Detect guarded string literal syntax
+    ///
+    /// RFC 3598 reserved this syntax for future use. As of Rust 2024,
+    /// using this syntax produces an error. In earlier editions, however, it
+    /// only results in an (allowed by default) lint, and is treated as
+    /// separate tokens.
+    fn report_guarded_str(&mut self, start: BytePos, str_before: &'src str) -> TokenKind {
+        let span = self.mk_sp(start, self.pos);
+        let edition2024 = span.edition().at_least_rust_2024();
+
+        // Prior to edition 2024, check the previous char, so we don't
+        // lint for each `#` before and after the string.
+        if !edition2024
+            && start > self.start_pos
+            && matches!(self.src.as_bytes()[self.src_index(start) - 1], b'#' | b'"')
+        {
+            return self.split_guarded_str_maybe(start, str_before);
```

Comment:
This doesn't look like a common situation and the early return shouldn't be important for performance.
So it would be simpler to wrap the skipping condition around the `buffer_lint` below (and then inline and remove `fn split_guarded_str_maybe`).